File: app/routers/story.py
import json
from fastapi import APIRouter, Request, Depends
from fastapi.responses import StreamingResponse
from sqlalchemy.ext.asyncio import AsyncSession
from app.core.database import get_db
from app.schemas.story import AnalyzeIdeaRequest, GenerateOutlineRequest, GenerateScriptRequest, ChatRequest, RefineRequest, WorldBuildingStartRequest, WorldBuildingTurnRequest
from app.services.story_llm import analyze_idea, generate_outline, generate_script, chat, refine, world_building_start, world_building_turn
from app.services import story_repository as repo

router = APIRouter(prefix="/api/v1/story", tags=["story"])


def get_llm_config(request: Request):
    return {
        "api_key": request.headers.get("X-LLM-API-Key", ""),
        "base_url": request.headers.get("X-LLM-Base-URL", ""),
        "provider": request.headers.get("X-LLM-Provider", ""),
    }


@router.post("/analyze-idea")
async def api_analyze_idea(req: AnalyzeIdeaRequest, request: Request, db: AsyncSession = Depends(get_db)):
    return await analyze_idea(req.idea, req.genre, req.tone, db=db, **get_llm_config(request))


@router.post("/generate-outline")
async def api_generate_outline(req: GenerateOutlineRequest, request: Request, db: AsyncSession = Depends(get_db)):
    return await generate_outline(req.story_id, req.selected_setting, db=db, **get_llm_config(request))


@router.post("/chat")
async def api_chat(req: ChatRequest, request: Request, db: AsyncSession = Depends(get_db)):
    cfg = get_llm_config(request)

    async def event_stream():
        try:
            async for chunk in chat(req.story_id, req.message, db=db, **cfg):
                yield f"data: {chunk}\n\n"
        except Exception as e:
            yield f"data: [ERROR] {str(e)}\n\n"
        yield "data: [DONE]\n\n"

    return StreamingResponse(event_stream(), media_type="text/event-stream")


@router.post("/generate-script")
async def api_generate_script(req: GenerateScriptRequest, request: Request, db: AsyncSession = Depends(get_db)):
    cfg = get_llm_config(request)

    async def event_stream():
        scenes = []
        try:
            async for scene in generate_script(req.story_id, db=db, **cfg):
                if "__usage__" not in scene:
                    scenes.append(scene)
                    yield f"data: {json.dumps(scene, ensure_ascii=False)}\n\n"
                else:
                    yield f"data: {json.dumps(scene, ensure_ascii=False)}\n\n"
        except Exception as e:
            yield f"data: [ERROR] {str(e)}\n\n"
        # 保存完整剧本供第二阶段使用
        await repo.save_story(db, req.story_id, {"scenes": scenes})
        yield "data: [DONE]\n\n"

    return StreamingResponse(event_stream(), media_type="text/event-stream")


@router.post("/refine")
async def api_refine(req: RefineRequest, request: Request, db: AsyncSession = Depends(get_db)):
    return await refine(req.story_id, req.change_type, req.change_summary, db=db, **get_llm_config(request))


@router.post("/world-building/start")
async def api_wb_start(req: WorldBuildingStartRequest, request: Request, db: AsyncSession = Depends(get_db)):
    return await world_building_start(req.idea, db=db, **get_llm_config(request))


@router.post("/world-building/turn")
async def api_wb_turn(req: WorldBuildingTurnRequest, request: Request, db: AsyncSession = Depends(get_db)):
    return await world_building_turn(req.story_id, req.answer, db=db, **get_llm_config(request))


@router.post("/{story_id}/finalize")
async def finalize_script(story_id: str, db: AsyncSession = Depends(get_db)):
    """把第一阶段剧本序列化为文本，供第二阶段 pipeline 使用"""
    story = await repo.get_story(db, story_id)
    scenes = story.get("scenes", [])
    if not scenes:
        from fastapi import HTTPException
        raise HTTPException(status_code=404, detail="剧本尚未生成，请先调用 generate-script")

    lines = []
    for ep in scenes:
        lines.append(f"# 第{ep['episode']}集 {ep['title']}")
        for s in ep.get("scenes", []):
            lines.append(f"\n## 场景{s['scene_number']}")
            lines.append(f"【环境】{s['environment']}")
            lines.append(f"【画面】{s['visual']}")
            for a in s.get("audio", []):
                lines.append(f"【{a['character']}】{a['line']}")

    script_text = "\n".join(lines)
    return {"story_id": story_id, "script": script_text}

File: frontend/src/api/story.js
import { useSettingsStore } from '../stores/settings.js'

function getHeaders() {
  const settings = useSettingsStore()
  const headers = { 'Content-Type': 'application/json' }
  if (settings.apiKey) headers['X-LLM-API-Key'] = settings.apiKey
  if (settings.llmBaseUrl) headers['X-LLM-Base-URL'] = settings.llmBaseUrl
  if (settings.provider) headers['X-LLM-Provider'] = settings.provider
  return headers
}

function getUrl(path) {
  const settings = useSettingsStore()
  const base = settings.backendUrl ? settings.backendUrl.replace(/\/$/, '') : ''
  return `${base}/api/v1/story${path}`
}

function getPipelineUrl(path) {
  const settings = useSettingsStore()
  const base = settings.backendUrl ? settings.backendUrl.replace(/\/$/, '') : ''
  return `${base}/api/v1/pipeline${path}`
}

export async function finalizeScript(storyId) {
  const res = await fetch(getUrl(`/${storyId}/finalize`), { method: 'POST', headers: getHeaders() })
  if (!res.ok) throw new Error(`请求失败 (${res.status})`)
  return res.json()
}

export async function startStoryboard(storyId, script, provider) {
  const url = getPipelineUrl(`/${storyId}/storyboard?script=${encodeURIComponent(script)}&provider=${provider}`)
  const res = await fetch(url, { method: 'POST', headers: getHeaders() })
  if (!res.ok) throw new Error(`请求失败 (${res.status})`)
  return res.json()
}

export async function getPipelineStatus(storyId) {
  const res = await fetch(getPipelineUrl(`/${storyId}/status`), { headers: getHeaders() })
  if (!res.ok) throw new Error(`请求失败 (${res.status})`)
  return res.json()
}

export async function analyzeIdea(idea, genre, tone) {
  const res = await fetch(getUrl('/analyze-idea'), {
    method: 'POST',
    headers: getHeaders(),
    body: JSON.stringify({ idea, genre, tone }),
  })
  if (!res.ok) throw new Error(`请求失败 (${res.status})`)
  return res.json()
}

export async function worldBuildingStart(idea) {
  const res = await fetch(getUrl('/world-building/start'), {
    method: 'POST',
    headers: getHeaders(),
    body: JSON.stringify({ idea }),
  })
  if (!res.ok) throw new Error(`请求失败 (${res.status})`)
  return res.json()
}

export async function worldBuildingTurn(storyId, answer) {
  const res = await fetch(getUrl('/world-building/turn'), {
    method: 'POST',
    headers: getHeaders(),
    body: JSON.stringify({ story_id: storyId, answer }),
  })
  if (!res.ok) throw new Error(`请求失败 (${res.status})`)
  return res.json()
}

export async function generateOutline(storyId, selectedSetting) {
  const res = await fetch(getUrl('/generate-outline'), {
    method: 'POST',
    headers: getHeaders(),
    body: JSON.stringify({ story_id: storyId, selected_setting: selectedSetting }),
  })
  if (!res.ok) throw new Error(`请求失败 (${res.status})`)
  return res.json()
}

export async function refineStory(storyId, changeType, changeSummary) {
  const res = await fetch(getUrl('/refine'), {
    method: 'POST',
    headers: getHeaders(),
    body: JSON.stringify({ story_id: storyId, change_type: changeType, change_summary: changeSummary }),
  })
  if (!res.ok) return null
  return res.json()
}

export async function streamChat(storyId, message, onChunk, onDone, onError) {
  let res
  try {
    res = await fetch(getUrl('/chat'), {
      method: 'POST',
      headers: getHeaders(),
      body: JSON.stringify({ story_id: storyId, message }),
    })
  } catch (e) {
    onError?.(e.message); return
  }
  if (!res.ok) { onError?.(`请求失败 (${res.status})`); return }

  const reader = res.body.getReader()
  const decoder = new TextDecoder()
  let buffer = ''

  try {
    while (true) {
      const { done, value } = await reader.read()
      if (done) break
      buffer += decoder.decode(value, { stream: true })
      const lines = buffer.split('\n')
      buffer = lines.pop()
      for (const line of lines) {
        if (!line.startsWith('data: ')) continue
        const payload = line.slice(6)
        if (payload === '[DONE]') { onDone(); return }
        if (payload.startsWith('[ERROR]')) { onError?.(payload.slice(8)); return }
        onChunk(payload)
      }
    }
  } catch (e) {
    onError?.(e.message); return
  }
  onDone()
}

export async function streamScript(storyId, onScene, onDone, onError) {
  let res
  try {
    res = await fetch(getUrl('/generate-script'), {
      method: 'POST',
      headers: getHeaders(),
      body: JSON.stringify({ story_id: storyId }),
    })
  } catch (e) {
    onError?.(e.message); return
  }
  if (!res.ok) { onError?.(`请求失败 (${res.status})`); return }

  const reader = res.body.getReader()
  const decoder = new TextDecoder()
  let buffer = ''

  try {
    while (true) {
      const { done, value } = await reader.read()
      if (done) break
      buffer += decoder.decode(value, { stream: true })
      const lines = buffer.split('\n')
      buffer = lines.pop()
      for (const line of lines) {
        if (!line.startsWith('data: ')) continue
        const payload = line.slice(6).trim()
        if (payload === '[DONE]') { onDone(); return }
        if (payload.startsWith('[ERROR]')) { onError?.(payload.slice(8).trim()); return }
        try { onScene(JSON.parse(payload)) } catch { onError?.('数据解析失败'); return }
      }
    }
    if (buffer.startsWith('data: ')) {
      const payload = buffer.slice(6).trim()
      if (payload && payload !== '[DONE]') {
        try { onScene(JSON.parse(payload)) } catch {}
      }
    }
  } catch (e) {
    onError?.(e.message); return
  }
  onDone()
}
